iroh-gossip/src/proto/hyparview.rs

```diff
@@ -0,0 +1,707 @@
+//! Implementation of the HyParView membership protocol
+//!
+//! The implementation is based on [this paper][paper] by Joao Leitao, Jose Pereira, Luıs Rodrigues
+//! and the [example implementation][impl] by Bartosz Sypytkowski
+//!
+//! [paper]: https://asc.di.fct.unl.pt/~jleitao/pdf/dsn07-leitao.pdf
+//! [impl]: https://gist.github.com/Horusiath/84fac596101b197da0546d1697580d99
+
+use std::{
+    collections::{HashMap, HashSet},
+    time::{Duration, Instant},
+};
+
+use derive_more::{From, Sub};
+use rand::rngs::ThreadRng;
+use rand::Rng;
+use serde::{Deserialize, Serialize};
+use tracing::debug;
+
+use super::{util::IndexSet, PeerAddress, PeerData, PeerInfo, IO};
+
+/// Input event for HyParView
+#[derive(Debug)]
+pub enum InEvent<PA> {
+    /// A [`Message`] was received from a peer.
+    RecvMessage(PA, Message<PA>),
+    /// A timer has expired.
+    TimerExpired(Timer<PA>),
+    /// A peer was disconnected on the IO layer.
+    PeerDisconnected(PA),
+    /// Send a join request to a peer.
+    RequestJoin(PA),
+    /// Update the peer data that is transmitted on join requests.
+    UpdatePeerData(PeerData),
+    /// Quit the swarm, informing peers about us leaving.
+    Quit,
+}
+
+/// Output event for HyParView
+#[derive(Debug)]
+pub enum OutEvent<PA> {
+    /// Ask the IO layer to send a [`Message`] to peer `PA`.
+    SendMessage(PA, Message<PA>),
+    /// Schedule a [`Timer`].
+    ScheduleTimer(Duration, Timer<PA>),
+    /// Ask the IO layer to close the connection to peer `PA`.
+    DisconnectPeer(PA),
+    /// Emit an [`Event`] to the application.
+    EmitEvent(Event<PA>),
+    /// New [`PeerData`] was received for peer `PA`.
+    PeerData(PA, PeerData),
+}
+
+/// Event emitted by the [`State`] to the application.
+#[derive(Clone, Debug)]
+pub enum Event<PA> {
+    /// A peer was added to our set of active connections.
+    NeighborUp(PA),
+    /// A peer was removed from our set of active connections.
+    NeighborDown(PA),
+}
+
+/// Kinds of timers HyParView needs to schedule.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum Timer<PA> {
+    DoShuffle,
+    PendingNeighborRequest(PA),
+}
+
+/// Messages that we can send and receive from peers within the topic.
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
+pub enum Message<PA> {
+    /// Sent to a peer if you want to join the swarm
+    Join(PeerData),
+    /// When receiving Join, ForwardJoin is forwarded to the peer's ActiveView to introduce the
+    /// new member.
+    ForwardJoin(ForwardJoin<PA>),
+    /// A shuffle request is sent occasionally to re-shuffle the PassiveView with contacts from
+    /// other peers.
+    Shuffle(Shuffle<PA>),
+    /// Peers reply to Shuffle requests with a random subset of their PassiveView.
```

Comment:
should mention both passive and active views
```suggestion
    /// Peers reply to [`Message::Shuffle`] requests with random peers from their passive and
    /// active views.
```